validate.go

```diff
@@ -0,0 +1,61 @@
+package cid
+
+import (
+	"fmt"
+
+	mh "github.com/multiformats/go-multihash"
+)
+
+var ErrPossiblyInsecureHashFunction = fmt.Errorf("potentially insecure hash functions not allowed")
+var ErrBelowMinimumHashLength = fmt.Errorf("hashes must be at least bytes long")
+
+const minimumHashLength = 20
+
+var goodset = map[uint64]bool{
+	mh.SHA2_256:     true,
+	mh.SHA2_512:     true,
+	mh.SHA3_224:     true,
+	mh.SHA3_256:     true,
+	mh.SHA3_384:     true,
+	mh.SHA3_512:     true,
+	mh.SHAKE_256:    true,
+	mh.DBL_SHA2_256: true,
+	mh.KECCAK_224:   true,
+	mh.KECCAK_256:   true,
+	mh.KECCAK_384:   true,
+	mh.KECCAK_512:   true,
+	mh.ID:           true,
+
+	mh.SHA1: true, // not really secure but still useful
```

Comment:
Could we restrict this to git only somehow?

Reply:
I don't necessary agree with this unless we are 100% sure SHA-1 are not used anywhere outside of git.  Otherwise we could break things, badly.

Reply:
The only reason we want to call SHA1 a "secure" hash because we want git compatibility. SHA1 is not a secure hash function. I *know* there are no other existing IPLD formats using SHA1 by default. If users have *chosen* to use SHA1 with IPLD, well, I'm not exactly sympathetic.

If we find some existing merkledag format that uses SHA1 and want to add compatibility, we can add that to the list later.

---

Note: I'm fine pushing this patch through as-is and fixing this later as this patch *is* a strict improvement.